### PR TITLE
Add llfio package

### PR DESCRIPTION
**Pull request by Arthapz**

### CODE REVIEW

File: packages/q/quickcpplib/xmake.lua [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+    add_urls("https://github.com/ned14/quickcpplib.git")
+    add_versions("20221116", "52163d5a198f1d0a2583e683f090778686f9f998")
+
+    add_deps("cmake")
+    on_install(function (package)
+        local configs = {}
+        table.insert(configs, "-DCMAKE_BUILD_TYPE=" .. (package:debug() and "Debug" or "Release"))
+        import("package.tools.cmake").install(package, configs)
+    end)
```
Comment (waruqi) [inline packages/q/quickcpplib/xmake.lua:15]: add a newline

File: packages/l/llfio/xmake.lua [hunk 2]
```diff
@@ -0,0 +1,25 @@
+package("llfio")
+    set_kind("library")
```
Comment (waruqi) [inline packages/l/llfio/xmake.lua:2]: it's default kind, please remove it.

File: packages/q/quickcpplib/xmake.lua [hunk 3]
```diff
@@ -0,0 +1,23 @@
+package("quickcpplib")
+    set_kind("library")
+    set_homepage("https://github.com/ned14/quickcpplib")
+    set_description("Eliminate all the tedious hassle when making state-of-the-art C++ 14 - 23 libraries!")
+    set_license("Apache 2")
```
Comment (waruqi) [inline packages/q/quickcpplib/xmake.lua:5]: Apache-2.0

File: packages/l/llfio/xmake.lua [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    set_homepage("https://github.com/ned14/llfio")
+    set_description("UTF8-CPP: UTF-8 with C++ in a Portable Way")
+    set_license("Apache 2")
```
Comment (waruqi) [inline packages/l/llfio/xmake.lua:5]: Apache-2.0

File: packages/q/quickcpplib/xmake.lua [hunk 5]
```diff
@@ -0,0 +1,23 @@
+package("quickcpplib")
+    set_kind("library")
```
Comment (waruqi) [inline packages/q/quickcpplib/xmake.lua:2]: it's default kind, please remove it.

File: packages/q/quickcpplib/xmake.lua [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    set_description("Eliminate all the tedious hassle when making state-of-the-art C++ 14 - 23 libraries!")
+    set_license("Apache 2")
+
+    add_urls("https://github.com/ned14/quickcpplib.git")
+    add_versions("20221116", "52163d5a198f1d0a2583e683f090778686f9f998")
+
+    add_deps("cmake")
+    on_install(function (package)
+        local configs = {}
+        table.insert(configs, "-DCMAKE_BUILD_TYPE=" .. (package:debug() and "Debug" or "Release"))
```
Comment (waruqi) [inline packages/q/quickcpplib/xmake.lua:13]: table.insert(configs, "-DBUILD_SHARED_LIBS=" .. (package:config("shared") and "ON" or "OFF"))

File: packages/l/llfio/xmake.lua [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+    add_urls("https://github.com/ned14/llfio/archive/refs/tags/$(version).tar.gz",
+             "https://github.com/ned14/llfio.git")
+    add_versions("all_tests_passed_ae7f9c5a92879285ad5100c89efc47ce1cb0031b", "a921d4c812fe3db98f0d4f7cb28c9fbdc5e2746a6b0283db5be03030557dcea6")
+
+    add_deps("cmake")
+    on_install(function (package)
+        local configs = {}
+        table.insert(configs, "-DCMAKE_BUILD_TYPE=" .. (package:debug() and "Debug" or "Release"))
```
Comment (waruqi) [inline packages/l/llfio/xmake.lua:14]: table.insert(configs, "-DBUILD_SHARED_LIBS=" .. (package:config("shared") and "ON" or "OFF"))

File: packages/l/llfio/xmake.lua [hunk 8]
```diff
@@ -0,0 +1,25 @@
+package("llfio")
+    set_kind("library")
+    set_homepage("https://github.com/ned14/llfio")
+    set_description("UTF8-CPP: UTF-8 with C++ in a Portable Way")
+    set_license("Apache 2")
+
+    add_urls("https://github.com/ned14/llfio/archive/refs/tags/$(version).tar.gz",
+             "https://github.com/ned14/llfio.git")
+    add_versions("all_tests_passed_ae7f9c5a92879285ad5100c89efc47ce1cb0031b", "a921d4c812fe3db98f0d4f7cb28c9fbdc5e2746a6b0283db5be03030557dcea6")
```
Comment (waruqi) [inline packages/l/llfio/xmake.lua:9]: incorrect git commit, it should be [ae7f9c5a92879285ad5100c89efc47ce1cb0031b](https://github.com/ned14/llfio/releases/tag/all_tests_passed_ae7f9c5a92879285ad5100c89efc47ce1cb0031b)

and invalid version number format, please use semver format.

```
    add_versions("2022.9.7", "ae7f9c5a92879285ad5100c89efc47ce1cb0031b")
```

File: packages/l/llfio/xmake.lua [hunk 9]
```diff
@@ -0,0 +1,25 @@
+package("llfio")
+    set_kind("library")
+    set_homepage("https://github.com/ned14/llfio")
+    set_description("UTF8-CPP: UTF-8 with C++ in a Portable Way")
+    set_license("Apache 2")
+
+    add_urls("https://github.com/ned14/llfio/archive/refs/tags/$(version).tar.gz",
+             "https://github.com/ned14/llfio.git")
```
Comment (waruqi) [inline packages/l/llfio/xmake.lua:8]:
```lua
    local versions = {
        ["2022.9.7"] = "ae7f9c5a92879285ad5100c89efc47ce1cb0031b"
    }
    add_urls("https://github.com/ned14/llfio/archive/refs/tags/all_tests_passed_$(version).tar.gz", {version = function (version)
        return versions[tostring(version)]
    end})
    add_urls("https://github.com/ned14/llfio.git")

    for version, commit in pairs(versions) do
        add_versions(version, commit)
    end
```

**Comment by waruqi** [comment]
https://github.com/xmake-io/xmake-repo/pull/1645

**Comment by Arthapz** [comment]
i was working on this :D juste taking my time because of some works

**Comment by Arthapz** [comment]
ned14 cmake is horrible, i will convert llfio and the needed libraries to XMake

File: packages/l/llfio/xmake.lua [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+    add_urls("https://github.com/ned14/llfio/archive/refs/tags/all_tests_passed_$(version).tar.gz", {version = function (version)
+        return hashes[tostring(version)]
+    end})
+    add_urls("https://github.com/ned14/llfio.git")
+
+    for version, commit in pairs(versions) do
+        add_versions(version, commit)
+    end
+
+    add_cxxflags("/EHsc")
```
[the remark this answers is not in the document]
Comment (Arthapz) [reply packages/l/llfio/xmake.lua:21]: yes i've done some things that i didn't pushed atm

**Comment by Arthapz** [comment]
Why my last commit is not referenced here

https://github.com/Arthapz/xmake-repo/commit/fc3a488d1719edb0e0a3f3c04f80346bac7b0c44

File: packages/l/llfio/xmake.lua [hunk 11]
```diff
@@ -14,40 +14,64 @@ package("llfio")
...
     end
 
-    add_cxxflags("/EHsc")
-    add_deps("quickcpplib", "outcome", "ntkernel-error-category")
+
+    add_deps("quickcpplib", "outcome", "ntkernel-error-category", "openssl")
+    on_load(function(package)
+        if package:config("headeronly") then
+            package:add("defines", "LLFIO_HEADERS_ONLY=1")
+            if is_plat("windows", "mingw") then
```
Comment (waruqi) [inline packages/l/llfio/xmake.lua:28]: package:is_plat

File: packages/l/llfio/xmake.lua [hunk 12]
```diff
@@ -14,40 +14,64 @@ package("llfio")
...
+                package:add("syslinks", "advapi32", "user32", "wsock32", "ws2_32", "ole32", "shell32")
+            end
+        else
+            if not package:config("shared") then
+                if is_plat("windows", "mingw") then
```
Comment (waruqi) [inline packages/l/llfio/xmake.lua:33]: and here

**Comment by waruqi** [comment]
maybe we can disable android platform.

**Comment by Arthapz** [comment]
std::filesystem on android is supported only from ndk 22, but the CI use ndk 21 so the compilation doesn't work :/

**Comment by waruqi** [comment]
> std::filesystem on android is supported only from ndk 22, but the CI use ndk 21 so the compilation doesn't work :/

you can switch ndk to 22 on ci

**Comment by Arthapz** [comment]
how ?

**Comment by Arthapz** [comment]
\o/ it work

**Comment by waruqi** [comment]
> \o/ it work

Great,but it still has compile errors on freebsd.

**Pull request merged**